src/libcore/borrow.rs

```diff
@@ -12,26 +12,157 @@
 
 #![stable(feature = "rust1", since = "1.0.0")]
 
-/// A trait for borrowing data.
+/// A trait identifying how borrowed data behaves.
 ///
-/// In general, there may be several ways to "borrow" a piece of data.  The
-/// typical ways of borrowing a type `T` are `&T` (a shared borrow) and `&mut T`
-/// (a mutable borrow). But types like `Vec<T>` provide additional kinds of
-/// borrows: the borrowed slices `&[T]` and `&mut [T]`.
+/// In Rust, it is common to provide different representations of a type for
+/// different use cases. For instance, storage location and management for a
+/// value can be specifically chosen as appropriate for a particular use via
+/// pointer types such as [`Box<T>`] or [`Rc<T>`] or one can opt into
+/// concurrency via synchronization types such as [`Mutex<T>`], avoiding the
+/// associated cost when in parallel doesn’t happen. Beyond these generic
+/// wrappers that can be used with any type, some types provide optional
+/// facets providing potentially costly functionality. An example for such a
+/// type is [`String`] which adds the ability to extend a string to the basic
+/// [`str`]. This requires keeping additional information unnecessary for a
+/// simple, imutable string.
 ///
-/// When writing generic code, it is often desirable to abstract over all ways
-/// of borrowing data from a given type. That is the role of the `Borrow`
-/// trait: if `T: Borrow<U>`, then `&U` can be borrowed from `&T`.  A given
-/// type can be borrowed as multiple different types. In particular, `Vec<T>:
-/// Borrow<Vec<T>>` and `Vec<T>: Borrow<[T]>`.
+/// These types signal that they are a specialized representation of a basic
+/// type `T` by implementing `Borrow<T>`. The method `borrow` provides a way
+/// to convert a reference to the type into a reference to the underlying
+/// basic type.
```

Comment:
This description sounds more like `Deref` than `Borrow`. I would say `Deref` is "a way to convert a reference to the type into a reference to the underlying basic type." The major difference between `Deref` and `Borrow` comes from the phrasing "**the** underlying basic type." A type gets at most one `Deref` impl so there must be **one** underlying basic type. I think the existing `Borrow` documentation handles this well where it discusses how there are some types that can be borrowed as multiple different types.